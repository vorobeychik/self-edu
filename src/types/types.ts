export type VideoType = {
  id: string,
  snippet:{
    title: string,
    channelId: string,
    channelTitle: string,
    thumbnails:{
      high:{
        url: string,
      }
      standard:{
        url: string,
      }
    }
  },
  statistics:{
    viewCount: number,
    likeCount: number,
  },
  channel:Channel,
};

export type Channel = {
  id: string,
  snippet:{
    title: string,
    thumbnails:{
      default:{
        url: string,
      },
      medium:{
        url: string,
      },
    }
  }
};
export type VideoId = {
  'kind': string,
  'etag': string,
  'id': {
    'kind': string,
    'videoId': string
  }
};