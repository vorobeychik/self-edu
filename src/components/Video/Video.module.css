a{
    text-decoration: none;
}

.video{
    width: calc(100% / var(--amount) - 20px);
    cursor: pointer;
    margin: 10px;

}

.channel_icon{
    width: 36px;
    height: 36px;
    border-radius: 45%;
    margin-right: 10px;
    margin-top: 50%;
}

.video_description{
    display: flex;
}

.video_preview{
    width: 100%;
    height: 65%;
}

.video_title{
    font-weight: bold;
    text-overflow: ellipsis;
    overflow: hidden;
    line-height: 1.4rem;
    font-size: 1rem;
    max-height: 4rem;
    display: -webkit-box;
    -webkit-line-clamp: 3;
    -webkit-box-orient: vertical;
    text-decoration: none;
    color: black;
}

.video_channel_name{
    color: dimgray;
    font-size:  0.9rem;
    font-weight: 400;
}

.video_views{
    color: dimgray;
    font-size:  0.8rem;
    font-weight: 400;
}

.video_likes{
    color: dimgray;
    font-size:  0.8rem;
    font-weight: 400;
    margin-left: 10px;
}

.video_statistic_container{
    display: flex;
}

.video_description_text{
    padding-right: 0.5rem;
}

@media (max-width: 1100px) and (min-width: 400px){
    .video_statistic_container {
        flex-direction: column;
    }
    .video_likes{
        margin: 0;
    }
}
@media (max-width: 1400px) {
    .video_title{
        font-size: 0.9rem;
    }


}

@media (max-width: 1880px) {
    .video_title{
        font-size: 0.8rem;
    }
}
