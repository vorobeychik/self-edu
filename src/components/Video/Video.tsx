import React, { useState } from 'react';
import styles from './Video.module.css';
import Skeleton from 'react-loading-skeleton';
import { VideoType } from '../../types/types';

interface VideoProps{
  video: VideoType,
}

const Video = ({ video }:VideoProps) => {
  let [isPreviewLoaded, setLoadedPreview] = useState(false);
  let [isIconLoaded, setLoadedIcon] = useState(false);

  function previewLoadHandler() {
    setLoadedPreview(true);
  }

  function iconLoadHandler(){
    setLoadedIcon(true);
  }


  return (
        <div className={styles.video} data-testid="video">
            <a href={`https://www.youtube.com/watch?v=${video.id}`}>
                <img
                    className={styles.video_preview}
                    src={video.snippet.thumbnails.high.url}
                    onLoad={previewLoadHandler}
                    style={{ display: isPreviewLoaded ?  'initial' : 'none'  }}
                />
            </a>
            {!isPreviewLoaded && <Skeleton height={195} width={'100%'} />}
            <div className={styles.video_description}>
                <a href={'https://www.youtube.com/channel/UCz7ttPTNO_S1PS9rNsCt_iA'}>
                    <img
                        src={ video.channel.snippet.thumbnails.medium.url}
                        className={styles.channel_icon}
                        onLoad={iconLoadHandler}
                        onError={iconLoadHandler}
                        style={{ display: isIconLoaded ?  'initial' : 'none'  }}
                    />
                </a>
                {!isIconLoaded && <Skeleton circle={true} height={36} width={36} style={{ marginRight: '10px',
                  marginTop: '50%' }}/>}
                <div className={styles.video_description_text}>
                    <a href={`https://www.youtube.com/watch?v=${video.id}`}>
                        <p className={styles.video_title}>{ video.snippet.title }</p>
                    </a>
                    <p className={styles.video_channel_name}>{ video.snippet.channelTitle }</p>
                    <div className={styles.video_statistic_container}>
                        <p className={styles.video_views}>Views { video.statistics.viewCount }</p>
                        <p className={styles.video_likes}>Likes { video.statistics.likeCount }</p>
                    </div>
                </div>
            </div>
        </div>
  );
};

export default Video;