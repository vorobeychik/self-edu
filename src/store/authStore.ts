import { makeAutoObservable } from 'mobx';

export class Auth{

  isAuth = false;

  constructor() {
    makeAutoObservable(this);
  }

  setAuth( value:boolean ){
    this.isAuth = value;
  }
}

