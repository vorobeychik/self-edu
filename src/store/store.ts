import { autorun, makeAutoObservable } from 'mobx';
import { initialState } from '../constants/const';
import {
  Board, Category, Comments, Note, PlaceHolder, Property, Tag, User,
} from '../types/types';
import {
  createBoard, userAuth, updateBoard, deleteBoard,
} from '../requests/requests';
import {
  createBoardTemplate, createCalendarData,
  createCategoryTemplate, createComment,
  createNoteTemplate,
  createPropertyTemplate, createTagTemplate,
  deleteAllCookies, getMonth,
  insert, isUserUsePhone,
} from '../utils/utils';

export class Store {
  user: User = initialState;

  isLoading = false;

  selectedNote: null | Note = null;

  selectedBoardId: string = '';

  isNoteOpen = false;

  searchedQuery = '';

  tagFilter: Tag | null = null;

  calendar = createCalendarData();

  isUserUsePhone = isUserUsePhone();

  constructor() {
    makeAutoObservable(this, {

    }, {
      deep: true,
      autoBind: true,
    });
  }

  selectBoard(boardId: string) {
    this.tagFilter = null;
    this.selectedBoardId = boardId;
  }

  setSearchQuery(query: string) {
    this.searchedQuery = query;
  }

  get searchedCategories() {
    if (this.selectedBoard) {
      return this.selectedBoard.categories
        .map((category) => ({
          ...category,
          notes: category.notes.filter((note) => new RegExp(this.searchedQuery).test(note.title)),
        }));
    }
  }

  get selectedMonth() {
    const month = getMonth();
    return this.calendar[0].months[month];
  }

  get notesFilterdByTag() {
    if (this.tagFilter && this.selectedBoard) {
      return this.selectedBoard.categories
        .map((category) => ({
          ...category,
          notes: category
            .notes.filter((note) => note.tags.some((element) => element.id === this.tagFilter!.id)),
        }));
    }
  }

  get selectedBoard() {
    return this.user.boards.find((board: Board) => board.id === this.selectedBoardId);
  }

  get selectedCategories() {
    if (this.tagFilter) {
      return this.notesFilterdByTag;
    } if (this.searchedQuery) {
      return this.searchedCategories;
    }
    return this.selectedBoard!.categories;
  }

  get isAuth() {
    return !!this.user.login;
  }

  get tags() {
    const tags: Tag[] = [];
    if (this.selectedBoard) {
      this.selectedBoard.categories.forEach((category) => {
        category.notes.forEach((note) => {
          note.tags.forEach((tag) => {
            if (tags.every((element) => element.id !== tag.id)) {
              tags.push(tag);
            }
          });
        });
      });
      return tags;
    }
  }

  openNote() {
    this.isNoteOpen = true;
  }

  closeNote() {
    this.isNoteOpen = false;
  }

  logOut() {
    this.user = initialState;
    deleteAllCookies();
  }

  async deleteCategory(category: Category) {
    if (this.selectedBoard) {
      this.selectedBoard.categories = this.selectedBoard.categories
        .filter((categoryElement) => category.id !== categoryElement.id);
    }
  }

  setTegFilter(tag: Tag) {
    this.tagFilter = tag;
  }

  unselectTagFilter() {
    this.tagFilter = null;
  }

  async changeNotePlace(placeHolder: PlaceHolder) {
    const selectedCategory = this.selectedBoard!.categories
      .find((category) => category.id === placeHolder.categoryId);
    const opendCategory = this.selectedBoard!.categories
      .find((category) => category.notes.some((el) => el.id === this.selectedNote?.id));

    if (selectedCategory && opendCategory && this.selectedNote) {
      this.selectedNote.status = selectedCategory.name;
      opendCategory.notes = opendCategory.notes.filter((elementNote) => elementNote.id !== this.selectedNote!.id);
      selectedCategory.notes = insert(selectedCategory.notes, placeHolder.index, this.selectedNote);
    }
  }

  async createNewBoard() {
    const board = createBoardTemplate('Goals');
    this.user.boards.push(board);
    await createBoard(board, this.user._id);
  }

  async createNewNote(category: Category) {
    const selectedCategory = this.selectedBoard!.categories.find((categoryElement) => category.id === categoryElement.id);
    if (selectedCategory) {
      selectedCategory.notes.push(createNoteTemplate(category.name));
    }
  }

  async addProperty() {
    if (this.selectedNote) {
      this.selectedNote.properties.push(createPropertyTemplate());
    }
  }

  async deleteProperty(property: Property) {
    if (this.selectedNote) {
      this.selectedNote.properties = this.selectedNote.properties.filter((propertyElement) => property.id !== propertyElement.id);
    }
  }

  async addComment(note: Note, title: string) {
    if (this.selectedNote) {
      this.selectedNote.comments.push(createComment(title));
    }
  }

  async deleteComment(comment: Comments) {
    if (this.selectedNote && this.selectedBoard) {
      this.selectedNote.comments = this.selectedNote.comments.filter((commentElement) => comment.id !== commentElement.id);
    }
  }

  async addTag(title: string) {
    if (this.selectedBoard && this.selectedNote && this.tags) {
      const tag = this.tags.find((element) => element.title === title);
      this.selectedNote.tags.push(tag ?? createTagTemplate(title));
    }
  }

  async deleteTag(tag: Tag) {
    if (this.selectedNote && this.selectedBoard) {
      this.selectedNote.tags = this.selectedNote.tags.filter((tagElement) => tagElement.id !== tag.id);
    }
  }

  async changePropertyTitle(title: string, property: Property) {
    const selectedProperty = this.selectedNote?.properties.find(
      (propertyElement) => propertyElement.id === property.id,
    );
    if (this.selectedBoard && selectedProperty) {
      selectedProperty.title = title;
    }
  }

  async changePropertyValue(value: string, property: Property) {
    const selectedProperty = this.selectedNote?.properties.find(
      (propertyElement) => propertyElement.id === property.id,
    );
    if (this.selectedBoard && selectedProperty) {
      selectedProperty.value = value;
    }
  }

  async deleteNote(note: Note) {
    const category = this.selectedBoard!.categories.find(
      (categoryElement) => categoryElement.notes.some((el) => el.id === note.id),
    );
    if (category && this.selectedBoard) {
      category.notes = category.notes.filter((el) => el.id !== note.id);
    }
  }

  async updateCategoryTitle(title: string, category: Category) {
    const selectedCategory = this.selectedBoard!.categories.find(
      (categoryElement) => category.id === categoryElement.id,
    );
    if (this.selectedBoard && selectedCategory) {
      selectedCategory.name = title;
      selectedCategory.notes = selectedCategory.notes.map((note) => ({ ...note, status: title }));
    }
  }

  async createNewCategory() {
    if (this.selectedBoard) {
      this.selectedBoard.categories.push(createCategoryTemplate());
    }
  }

  async updateBoardTitle(title: string) {
    if (this.selectedBoard) {
      this.selectedBoard.name = title;
    }
  }

  async updateNoteTitle(title: string) {
    if (this.selectedNote && this.selectedBoard) {
      this.selectedNote.title = title;
    }
  }

  async authorize() {
    this.isLoading = true;
    const user = await userAuth();
    this.isLoading = false;
    if (user) {
      this.selectedBoardId = user.boards[0].id;
      this.user = user;
    }
  }

  async deleteBoard(board: Board) {
    if (this.selectedBoard) {
      this.user.boards = this.user.boards.filter((boardElement) => board.id !== boardElement.id);
      this.selectedBoardId = this.user.boards[0].id;

      await deleteBoard(this.user._id, board.id);
    }
  }

  selectNote(note: Note) {
    this.selectedNote = note;
  }

  unselectNote() {
    this.selectedNote = null;
  }
}
export const store = new Store();

autorun(async () => {
  if (store.selectedBoard) {
    await updateBoard(store.selectedBoard);
  }
});
