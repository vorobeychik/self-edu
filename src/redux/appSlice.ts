import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { getVideos } from '../requests/requests';
import { RootState } from './store';
import { VideoType } from '../types/types';




export const searchVideos:any = createAsyncThunk(
  'app/fetchVideos',
  async ({ query, searchMode }:{ query: string, searchMode: SearchMode }) => {
    const videos = await getVideos(query);
    return { videos, searchMode };
  },
);

type AppInitialState = {
  videos: VideoType[],
  query: string,
};

const initialState:AppInitialState = {
  query: '',
  videos: [],
};

export enum SearchMode{
  Add = 'Add',
  Insert = 'Insert',
}

const appSlice = createSlice({
  name: 'app',
  initialState,
  reducers:{
    setQuery:(state, action:PayloadAction<string>) =>{
      state.query = action.payload;
    },
  },
  extraReducers:{
    [searchVideos.fulfilled]: (state:AppInitialState, action:PayloadAction<{ videos:VideoType[], searchMode: SearchMode }> ) => {
      if (action.payload.searchMode === SearchMode.Insert){
        state.videos = action.payload.videos;
      } else {
        state.videos = [...state.videos, ... action.payload.videos];
      }

    },
  },
});

export default appSlice.reducer;
export const { setQuery  } = appSlice.actions;

export const selectQuery = (state:RootState) => state.app.query;
export const selectVideos = (state: RootState) => state.app.videos;