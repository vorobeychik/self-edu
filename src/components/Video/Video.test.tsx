import Video from './Video';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';
import { VideoType } from '../../types/types';

const video:VideoType = {
  id: '123123',
  snippet:{
    title: 'Js',
    channelId: '123123123',
    channelTitle: 'learn js',
    thumbnails:{
      high:{
        url: 'asd',
      },
      standard:{
        url: 'asd',
      },
    },
  },
  statistics:{
    viewCount: 123123,
    likeCount: 123123,
  },
  channel:{
    snippet:{
      title: 'learn js',
      thumbnails:{
        default:{
          url: 'asdasdasd',
        },
        medium:{
          url: 'asdasd',
        },
      },
    },
  },
};

describe('Video component', () => {
  it('should render a components', () => {
    const { getByTestId } = render(
          <Provider store={store}>
              <Video video={video} />
          </Provider>);


  });
});