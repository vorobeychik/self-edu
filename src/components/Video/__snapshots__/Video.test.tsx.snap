// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video component should render a components 1`] = `
<div
  data-testid="video"
>
  <a
    href="https://www.youtube.com/watch?v=123123"
  >
    <img
      src="asd"
      style="display: none;"
    />
  </a>
  <span>
    <span
      class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
      style="width: 100%; height: 195px;"
    >
      ‌
    </span>
  </span>
  <div>
    <a
      href="https://www.youtube.com/channel/UCz7ttPTNO_S1PS9rNsCt_iA"
    >
      <img
        src="asdasd"
        style="display: none;"
      />
    </a>
    <span>
      <span
        class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
        style="margin-right: 10px; margin-top: 50%; width: 36px; height: 36px; border-radius: 50%;"
      >
        ‌
      </span>
    </span>
    <div>
      <a
        href="https://www.youtube.com/watch?v=123123"
      >
        <p>
          Js
        </p>
      </a>
      <p>
        learn js
      </p>
      <div>
        <p>
          Views 
          123123
        </p>
        <p>
          Likes 
          123123
        </p>
      </div>
    </div>
  </div>
</div>
`;
